### ISSUE TITLE
Breaks with React-Redux v7

### ISSUE THREAD

**Issue by mongkuen** [electron-userland/electron-webpack#283]
React-Redux v7 started using hooks, and this somehow breaks the `connect` function when run through electron-webpack like so: 

<img width="1134" alt="Screen Shot 2019-04-28 at 7 18 42 PM" src="https://user-images.githubusercontent.com/5866832/56872094-cf23bb80-69ea-11e9-8463-5f2f43aa5913.png">

Issue is easily replicable and I've provided an example using the quickstart project [here](https://github.com/mongkuen/electron-webpack-react-redux)

I've verified the same react-redux code is verified to be working from within create-react-app's build system. 

React-redux v6 does not present this problem, as it uses context. However that meant it had significant performance issues over v7.

**Comment by loopmode** [comment]
The problem is in https://github.com/electron-userland/electron-webpack/blob/8b012241fe615b9fe7da956eaf6ad8ecdd1d018d/packages/electron-webpack/src/main.ts#L261

Using these snippets from the ["duplicate react" section](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react), you can verify that there are two versions of react being loaded:

```
// Add this in node_modules/react-dom/index.js
window.React1 = require('react');

// Add this in your component file
require('react-dom');
window.React2 = require('react');
console.log(window.React1 === window.React2);
```

It does log `false` - because react is added to the "whiteListedModules" and somehow treated differently. I'm not quite sure what the whitelisting mechanism does, but once you comment out the line in `node_modules/electron-webpack/out/main.js`, your example works again.

So, we need to evaluate a couple of things here, like what exactly does whitelisting do? And is it really necessary to implicitly add react to those whitelisted modules? 
It obviously breaks the hooks system provided by latest react versions - so it's a rather critical issue.

I'll investigate a little more but I'm not sure how much time I can invest in this.

**Comment by loopmode** [comment]
Okay. So... there is a very simple "solution" - but it really is a workaround in my opinion, and it feels fragile.

You just override the `externals` object with an empty one, and the problem is gone.

Provide a [configuration](https://webpack.electron.build/configuration) for electron-webpack, define a webpackConfig for the renderer, and in that config override the `externals`.

For example, add this to `package.json`:
```
"electronWebpack": {
    "renderer": {
      "webpackConfig": "webpack.renderer.js"
    }
  }
```

Then add a file `webpack.renderer.js` with this content:
```
module.exports = {
  externals: {}
};
```

Now it should work. I logged the final config in the console, and I got:
```
┏ Renderer -------------------

  { config:
     { context: 'D:\\Projects\\@github.com\\electron-webpack-react-redux',
       devtool: 'eval-source-map',
       externals: {},
...
```

So the `externals` actually get flushed and it works. However, since currently `webpack-merge` is used with its "smart" strategy (hard-coded), I was surprised to see this work. I didn't expect the empty object to override the existing one. This is the same log without overriding the `externals` value:
```
┏ Renderer -------------------

  { config:
     { context: 'D:\\Projects\\@github.com\\electron-webpack-react-redux',
       devtool: 'eval-source-map',
       externals:
        [ 'react-redux',
          'redux',
          'source-map-support',
          'electron',
          'webpack',
          'electron-devtools-installer' ],
```

So, as you can see, we remove several other externals - not sure about consequences (e.g. performance?)


Long story short:
- For now, you can override externals and "make it work"
- However, `electron-webpack` should probably just not automatically add react to the externals

**Comment by loopmode** [comment]
Okay, after some further research, the solution is rather simple, as somewhat documented [here](https://webpack.electron.build/configuration#white-listing-externals).

All you need to do is add `react-redux` to the whitelist.

Add this to your package.json:

```
  "electronWebpack": {
    "whiteListedModules": [
      "react-redux"
    ]
  },
```

**Comment by mongkuen** [comment]
That's incredibly helpful, thank you!

**Comment by 0xhjohnson** [comment]
@loopmode ran into this exact issue, your whitelist solution works fine for both react-redux and react-useportal

**Comment by tmacwill** [comment]
just ran into this as well, thanks for posting the solution :)

is it possible to make this the default behavior?

**Comment by mrpierrot** [comment]
You save my day!!!!

**Comment by andreibursuc97** [comment]
may god bless you @loopmode

**Comment by loopmode** [comment]
I wish we could change the default behavior with those whitelisted modules, but it would sure be a breaking change for many users. I believe the idea was to make _all_ dependencies "webpack externals" by default with the assumption: it's not the web, it's electron, we don't need to actually bundle up everything. Well, it turns out the assumption was incorrect and it is only true for dependencies you use on the main process of an electron application, but not for any dependencies you use in the renderer process.

(Reminder: whitelisting in electron-webpack means _not_ putting a dependency to webpack externals) 

And.. naively and optimistically imagining a future.. maybe we could actually make a really new version that is based on workspaces, where renderer and main are two actually separate packages with their own sets of dependencies.. then we could make it work out of the box for everyone by putting all dependencies of the main process package into webpack externals, and none of the renderer package dependencies, still allowing users to do it manually.

However, I fear that we (as in: the electron-webpack developer community) are just not active and motivated enough to actually pull it off. @develar did such a great job putting up this great project, but the rest of us is just too busy doing our own stuff or just using his/this stuff and nobody takes initiative, sadly. I personally haven't worked on an actual electron project in years... 

Sorry for the unasked-for monologue :)
And I hope somebody proves me wrong and goes ahead with a totally awesome v3.0 branch ;)
